### PR TITLE
storage: add a new system var for the future new persist_sink

### PR DESCRIPTION
**Pull request by guswynn**
This pr is broken out of https://github.com/MaterializeInc/materialize/pull/17589, and only adds the system parameter (which will later be connected to a LaunchDarkly flag), that controls the new storage `persist_sink`. This is because this was non-trival:

- A new `InternalCommand::UpdateConfiguration`, which is required so different workers don't setup dataflows differently
- Some subtle changes to the persist sink metrics handling, so we can use a metric to track which impl a source is using
- Various other additions to system var handling, as this is the first non-persist storage var being added

### Motivation

  * This PR adds a known-desirable feature.

Part of https://github.com/MaterializeInc/materialize/pull/17589

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
  - Tested locally. note that this pr doesn't actually add any new functionality! 
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
https://github.com/MaterializeInc/materialize/blob/6b03c7d6fa75bc9ea1c7e8608a2585da818c3e02/doc/developer/design/20230210_storage_persist_sink_redux.md
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

### CODE REVIEW

**Review by jkosh44 — APPROVED** [review]
surfaces changes LGTM

File: src/adapter/src/session/vars.rs [hunk 1]
```diff
@@ -1141,6 +1153,9 @@ pub struct SystemVars {
     // features
     enable_with_mutually_recursive: SystemVar<bool>,
 
+    // storage configuration
+    enable_multi_worker_storage_persist_sink: SystemVar<bool>,
```
Comment (aalexandrov) [inline src/adapter/src/session/vars.rs:1157]: If you don't want this to be accidentally overridden by a LD parameter you should also add the following line to `SystemVars::iter_synced()`:

```rust
.filter(|v| v.name() != ENABLE_MULTI_WORKER_STORAGE_PERSIST_SINK.name)
```

File: src/adapter/src/session/vars.rs [hunk 2]
```diff
@@ -379,6 +379,18 @@ const PERSIST_COMPACTION_MINIMUM_TIMEOUT: ServerVar<Duration> = ServerVar {
     safe: true,
 };
 
+/// Controls whether or not to use the new storage `persist_sink` implementation in storage
+/// ingestions.
+const ENABLE_MULTI_WORKER_STORAGE_PERSIST_SINK: ServerVar<bool> = ServerVar {
```
Comment (aalexandrov) [inline src/adapter/src/session/vars.rs:384]: Nit: isn't `ENABLE_MULTI_WORKER_PERSIST_SINK` or `ENABLE_MULTI_WORKER_STORAGE_SINK` equally descriptive? Do we differentiate between storage persist sinks and dataflow persist sinks?

File: src/adapter/src/session/vars.rs (hunk 1 above)
Comment (guswynn) [reply src/adapter/src/session/vars.rs:1157]: I wan this to be controllable with and LD parameter tho!

File: src/adapter/src/session/vars.rs (hunk 2 above)
Comment (guswynn) [reply src/adapter/src/session/vars.rs:384]: there are:
- storage sinks (CREATE SINK)
- storage persist sinks (used for sources)
- compute persist sinks (used for MV's)

and this only applies to the first one!

File: src/storage-client/src/types/parameters.proto [hunk 3]
```diff
@@ -15,5 +15,6 @@ import "persist-client/src/cfg.proto";
 package mz_storage_client.types.parameters;
 
 message ProtoStorageParameters {
+    bool enable_multi_worker_storage_persist_sink = 2;
```
Comment (petrosagg) [inline src/storage-client/src/types/parameters.proto:18]: sorry for my ocd, can we keep the fields ordered by their ordinal?

File: src/storage/src/source/types.rs [hunk 4]
```diff
@@ -351,60 +351,75 @@ pub struct SourcePersistSinkMetrics {
...
 }
 
 impl SourcePersistSinkMetrics {
-    /// Initialises source metrics for a given (source_id, worker_id)
+    /// Initialises source metrics used in the `persist_sink`.
     pub fn new(
         base: &SourceBaseMetrics,
-        shard_id: &mz_persist_client::ShardId,
         source_id: GlobalId,
+        parent_source_id: GlobalId,
```
Comment (petrosagg) [inline src/storage/src/source/types.rs:365]: What is a parent source?

**Review by petrosagg — COMMENTED** [review]
It looks like we don't (and won't) dynamically tear down and recreate dataflows if the flag changes after the fact. If that's the case then shouldn't the flag be part of the `CreateSource` command which avoids adding an new internal command and storing the flag in the state?

File: src/storage/src/source/types.rs (hunk 4 above)
Comment (guswynn) [reply src/storage/src/source/types.rs:365]: Its the top-level source when the source is a subsource!

**Comment by guswynn** [comment]
> It looks like we don't (and won't) dynamically tear down and recreate dataflows if the flag changes after the fact. If that's the case then shouldn't the flag be part of the CreateSource command which avoids adding an new internal command and storing the flag in the state?

I am avoiding this so we don't maintain any additional infrastructure to set this up!

File: src/adapter/src/session/vars.rs (hunk 1 above)
Comment (aalexandrov) [reply src/adapter/src/session/vars.rs]: Sorry, than ignore this.

File: src/storage/src/internal_control.rs [hunk 5]
```diff
@@ -19,6 +19,18 @@ use mz_storage_client::controller::CollectionMetadata;
 use mz_storage_client::types::sinks::{MetadataFilled, StorageSinkDesc};
 use mz_storage_client::types::sources::IngestionDescription;
 
+/// Storage instance configuration parameters that can affect
+/// dataflow rendering, and as such must be applied to
+/// `StorageWorker`s in a consistent order with source and sink
```
Comment (aljoscha) [inline src/storage/src/internal_control.rs:24]: nit: but please ignore, I don't think there's a `StorageWorker` in our code

**Review by aljoscha — APPROVED** [review]
I think this is good to merge!

Some comments, also for other reviewers:
 - We _might_ want to restart dataflows based on configuration changes in the future, even though it's potentially costly (for example, rehydrating a UPSERT ingestion). I think it's okay to go with the approach that this only changes on restart for now.
 - Threading through the LD var/system var seems brittle. There seem to be a lot of `if else` branches where you have to manually make sure that you don't forget any?

**Comment by aalexandrov** [comment]
> I think it's okay to go with the approach that this only changes on restart for now.

Nit: make sure that the current behavior is reflected in the feature flag description in LaunchDarkly (see the `enable_with_mutually_recursive` for example).

**Pull request merged**